python/ray/train/callbacks/results_preprocessors/aggregate/aggregate_preprocessor.py

```diff
@@ -0,0 +1,150 @@
+import logging
+from typing import Dict, List, Optional
+
+import numpy as np
+
+from ray.util.debug import log_once
+from ray.util.annotations import DeveloperAPI
+from ray.train.callbacks.results_preprocessors.preprocessor import ResultsPreprocessor
+from ray.train.callbacks.results_preprocessors.aggregate.aggregate_fn import (
+    AggregateFn,
+    Average,
+    Max,
+    WeightedAverage,
+)
+from ray.train.callbacks.results_preprocessors.aggregate.aggregate_utils import (
+    VALID_AGGREGATE_TYPES,
+    _get_metrics_from_results,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class AggregateResultsPreprocessor(ResultsPreprocessor):
+    """A preprocessor that aggregates training metrics from all workers.
+
+    Args:
+        aggregation_fn (AggregateFn):
+            An aggregation method that performs the aggregation on results.
+        keys (Optional[List[str]]):
+            A list of keys reported in results to be aggregated. Keys should
+            be saved using ``train.report()``. If a key is invalid or not
+            reported, it will be ignored.
+    """
+
+    def __init__(self, aggregation_fn: AggregateFn, keys: Optional[List[str]] = None):
+        self.aggregate_fn = aggregation_fn
+        self.keys = keys
+
+    def preprocess(self, results: List[Dict]) -> List[Dict]:
+        """Aggregate results before sending them to callbacks.
+
+        Args:
+            results (List[Dict]): A list of results from all workers. The metrics
+                specified in ``keys`` will be averaged according by ``aggregation_fn``.
+                Non-numerical values will be ignored.
+        Returns:
+            An updated results list with aggregated results.
+        """
+        results = [] if results is None else results
+        if len(results) == 0:
+            return results
+
+        self.aggregate_fn.prepare(results)
+        reported_metrics = {key for result in results for key in result.keys()}
+
+        if self.keys is None:
+            valid_keys = []
+            for metric in reported_metrics:
```

Comment:
Can we move this validation check into `_get_metrics_from_results`? We are doing some validation in that method (checking if all workers have reported), but then some validation in this method (checking if the types are valid). Would be good to do all the validation in one place.

Reply:
Actually, nvm seems like we can just remove this. It's a duplicate of what's being done in `_get_metrics_from_results`.